### PR TITLE
OG-416: validate no extra bytes in msg.data

### PR DESCRIPTION
**Pull request by drortirosh**
- validate the packing of the relayCall request.
- make transaction penalizeable if the packing is not cannonicall, and
contains "rubbish" in excess of actual parameters (either between or
after parametrs)
- relayer now also make the test against the Penalizer, to avoid
submitting transactions that are penalizeable (either because of a bug
in the relayer, or malicious data from client)

### CODE REVIEW

File: contracts/utils/RelayHubValidator.sol [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import "../interfaces/GsnTypes.sol";
+library RelayHubValidator {
+
+    // validate that encoded relayCall is properly packed without any extra bytes
+    function verifyTransactionPacking(
+        GsnTypes.RelayRequest calldata relayRequest,
+        bytes calldata signature,
+        bytes calldata approvalData
+    ) internal pure {
+        uint expectedMsgDataLen = 4 + 22*32 +
```
Comment (forshtat) [inline contracts/utils/RelayHubValidator.sol:14]: please use explicit integer sizes

Comment (forshtat) [inline contracts/utils/RelayHubValidator.sol:14]: please define named constants

File: contracts/utils/RelayHubValidator.sol [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        len1(signature) + len1(approvalData) + len1(relayRequest.request.data) + len1(relayRequest.relayData.paymasterData);
+        require(signature.length <= 65, "invalid signature length");
+        require(expectedMsgDataLen == msg.data.length, "extra msg.data bytes" );
+    }
+
+    // helper method for verifyTransactionPacking
+    function len1(bytes calldata buf) internal pure returns (uint) {
```
Comment (forshtat) [inline contracts/utils/RelayHubValidator.sol:21]: please give meaningful names to methods

File: contracts/utils/RelayHubValidator.sol [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        return 32 + ((buf.length+31) & uint(~31));
```
Comment (forshtat) [inline contracts/utils/RelayHubValidator.sol:22]: please define named constants

File: test/RelayHub.test.ts [hunk 4]
```diff
@@ -422,6 +422,17 @@ contract('RelayHub', function ([_, relayOwner, relayManager, relayWorker, sender
           )
         })
 
+        it('should revert if encoded function contains extra bytes', async () => {
```
Comment (forshtat) [inline test/RelayHub.test.ts:425]: please add test for signature require

File: test/RelayHubValidator.test.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+const TestRelayHubValidator = artifacts.require('TestRelayHubValidator')
+
+contract('RelayHubValidator', ([from, senderAddress, target, paymaster, relayWorker, forwarder]) => {
+  let validator: any
+  before(async () => {
+    validator = await TestRelayHubValidator.new()
+  })
+
+  it('#len1 should return dynamic structure size', async () => {
+    const len1 = async function (size: number): Promise<number> { return validator.len1('0x' + '11'.repeat(size)).then((x: any) => x.toNumber()) }
```
Comment (forshtat) [inline test/RelayHubValidator.test.ts:12]: please give meaningful names to lambda functions

File: test/RelayHubValidator.test.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+const TestRelayHubValidator = artifacts.require('TestRelayHubValidator')
+
+contract('RelayHubValidator', ([from, senderAddress, target, paymaster, relayWorker, forwarder]) => {
+  let validator: any
+  before(async () => {
+    validator = await TestRelayHubValidator.new()
+  })
+
+  it('#len1 should return dynamic structure size', async () => {
```
Comment (forshtat) [inline test/RelayHubValidator.test.ts:11]: please start with 'should' at all costs

File: test/RelayHubValidator.test.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    { approvalData: '0x123456' },
+    { approvalData: '0x123456', signature: '0xabcd' },
+    { signature: '0xab' },
+    { data: '0x1234' },
+    { data: '0x1234', signature: '0xabcdef' },
+    { paymasterData: '0x12345678' },
+    { suffix: '123456' },
+    { data: '0x1234', signature: '0xabcdef', suffix: '123456' }
+  ].forEach((appended: any) => {
+    it(`test verify length with "${JSON.stringify(appended)}" `, async () => {
```
Comment (forshtat) [inline test/RelayHubValidator.test.ts:33]: please start with 'should' at all costs

**Review by forshtat — CHANGES_REQUESTED** [review]

File: contracts/utils/RelayHubValidator.sol (hunk 3 above)
Comment (drortirosh) [reply contracts/utils/RelayHubValidator.sol:22]: what name? `mask_ffffe0` ? I don't have a better name for this mask...

File: test/RelayHub.test.ts (hunk 4 above)
Comment (drortirosh) [reply test/RelayHub.test.ts:425]: there is, but it the Validator test. here is just a single test, just to verify the "verify" method is called.

File: test/RelayHubValidator.test.ts (hunk 6 above)
Comment (drortirosh) [reply test/RelayHubValidator.test.ts:11]: sorry, English no good for that. its a sanity check that the method does its work.
`#len1` should be enough, though

File: test/RelayHubValidator.test.ts (hunk 5 above)
Comment (drortirosh) [reply test/RelayHubValidator.test.ts:12]: localWrapperForSolidityLen1ToBeEasilyCalledFromAssert ?

File: contracts/utils/RelayHubValidator.sol (hunk 1 above)
Comment (drortirosh) [reply contracts/utils/RelayHubValidator.sol:14]: the old elaborated description was lost. added back.

File: test/RelayHub.test.ts (hunk 4 above)
Comment (forshtat) [reply test/RelayHub.test.ts:425]: there is no check for a `require('invalid signature length'`

File: test/RelayHubValidator.test.ts [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+            baseRelayFee: 2,
+            relayWorker,
+            paymaster: paymaster,
+            paymasterData: '0x',
+            clientId: 3,
+            forwarder
+          }
+        },
+        '0x' + '11'.repeat(66),
+        '0x', 0), 'invalid sig')
```
Comment (forshtat) [inline test/RelayHubValidator.test.ts:98]: "invalid signature length"

**Review by forshtat — APPROVED** [review]

**Pull request merged**